### PR TITLE
HHH-16472 Treated left join with restriction on entity results in incorrect query

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-16472

### CODE REVIEW

**Comment by beikov** [comment]
Depends on https://github.com/hibernate/hibernate-orm/pull/6484

**Comment by mbladel** [comment]
> Depends on #6484

Since that change is now merged I've rebased this PR and adapted the solution to work with the new treat handling.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java [hunk 1]
```diff
@@ -5000,19 +4999,38 @@ private Predicate consumeConjunctTreatTypeRestrictions(Map<TableGroup, Map<Strin
 					tableGroup,
 					this
 			);
+			// We need to check if this is a treated left or full join, which case we should
+			// allow null discriminator values to maintain correct semantics
+			final TableGroupJoin join = getParentTableGroupJoin( tableGroup );
+			final boolean allowNulls = join != null && ( join.getJoinType() == SqlAstJoinType.LEFT || join.getJoinType() == SqlAstJoinType.FULL );
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java:5005]: Updated the solution to allow nulless in the discriminator predicate (i.e. added `or t.disc_col is null`) when the treated table group is a `LEFT` or `FULL` join.

**Review by beikov — APPROVED** [review]
LGTM

**Pull request merged**